
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Part I: CrossChatH &#8212; CrossChat 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Part II: CrossChatT" href="CrossChatT_tutorial.html" />
    <link rel="prev" title="&lt;no title&gt;" href="index.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Part-I:-CrossChatH">
<h1>Part I: CrossChatH<a class="headerlink" href="#Part-I:-CrossChatH" title="Permalink to this heading">¶</a></h1>
<section id="Load-required-packages">
<h2>Load required packages<a class="headerlink" href="#Load-required-packages" title="Permalink to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">crosschat</span> <span class="k">as</span> <span class="nn">cc</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/umap/distances.py:1053: NumbaDeprecationWarning: <span class="ansi-bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/umap/distances.py:1061: NumbaDeprecationWarning: <span class="ansi-bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/umap/distances.py:1076: NumbaDeprecationWarning: <span class="ansi-bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/pynndescent/utils.py:202: NumbaDeprecationWarning: <span class="ansi-bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit(
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/pynndescent/utils.py:302: NumbaDeprecationWarning: <span class="ansi-bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit(
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/umap/umap_.py:646: NumbaDeprecationWarning: <span class="ansi-bold">The &#39;nopython&#39; keyword argument was not supplied to the &#39;numba.jit&#39; decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.</span>
  @numba.jit()
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/threadpoolctl.py:1019: RuntimeWarning: libc not found. The ctypes module in Python 3.8 is maybe too old for this OS.
  warnings.warn(
</pre></div></div>
</div>
</section>
<section id="Load-data-and-preprocess">
<h2>Load data and preprocess<a class="headerlink" href="#Load-data-and-preprocess" title="Permalink to this heading">¶</a></h2>
<p>You can choose to use your own hierarchical clustering result as input. Please load your clustering result as the format of user_comm_ids below.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># You can download the data from this link: https://drive.google.com/drive/folders/1NU8htAxVHrEiNV7YqSzEkHFGBnM6vNlM</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;../src/crosschat/data/covid3000.h5ad&#39;</span><span class="p">)</span>
<span class="c1"># user_comm_ids = np.load(&#39;../src/crosschat/data/user_comm_ids.npy&#39;)</span>
<span class="n">ccH_obj</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">CrossChatH</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">species</span><span class="o">=</span><span class="s2">&quot;human&quot;</span><span class="p">,</span><span class="n">user_comm_ids</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">prepare_adata</span><span class="p">(</span><span class="n">normalize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">input</span><span class="o">=</span><span class="s1">&#39;allgenes&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Perform-Multiscale-clustering">
<h2>Perform Multiscale clustering<a class="headerlink" href="#Perform-Multiscale-clustering" title="Permalink to this heading">¶</a></h2>
<p>To reduce noise, one can increase parameter values for <code class="docutils literal notranslate"><span class="pre">window_size</span></code> and <code class="docutils literal notranslate"><span class="pre">basin_radius</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1">#Change cluster_by to &quot;l&quot; or &quot;r&quot; if you want to cluster based on ligands/receptors only.</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">Multsicale_clustering</span><span class="p">(</span><span class="n">cluster_by</span><span class="o">=</span><span class="s2">&quot;allgenes&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 100/100 [00:02&lt;00:00, 49.32it/s]
100%|██████████| 100/100 [02:50&lt;00:00,  1.70s/it]
100%|██████████| 100/100 [00:15&lt;00:00,  6.66it/s]
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/threadpoolctl.py:1019: RuntimeWarning: libc not found. The ctypes module in Python 3.8 is maybe too old for this OS.
  warnings.warn(
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/threadpoolctl.py:1019: RuntimeWarning: libc not found. The ctypes module in Python 3.8 is maybe too old for this OS.
  warnings.warn(
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/threadpoolctl.py:1019: RuntimeWarning: libc not found. The ctypes module in Python 3.8 is maybe too old for this OS.
  warnings.warn(
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/threadpoolctl.py:1019: RuntimeWarning: libc not found. The ctypes module in Python 3.8 is maybe too old for this OS.
  warnings.warn(
</pre></div></div>
</div>
</section>
<section id="Draw-clustering-results">
<h2>Draw clustering results<a class="headerlink" href="#Draw-clustering-results" title="Permalink to this heading">¶</a></h2>
<p>You do not need to run select_partitions if you use your own hierarchical clustering.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">ccH_obj</span><span class="o">.</span><span class="n">select_partitions</span><span class="p">(</span><span class="n">window_size</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">basin_radius</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">lr</span> <span class="o">=</span> <span class="s2">&quot;L&quot;</span><span class="p">)</span>
<span class="c1"># If you would like to use your own hierarchical clustering, change cluster_input to &quot;userinput&quot;</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">Draw_multiscale_umap</span><span class="p">(</span><span class="n">cluster_input</span><span class="o">=</span><span class="s2">&quot;allgenes&quot;</span><span class="p">,</span><span class="n">spatial</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">Draw_annotations_umap</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_8_0.png" src="../_images/tutorials_CrossChatH_tutorial_8_0.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_8_1.png" src="../_images/tutorials_CrossChatH_tutorial_8_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_8_2.png" src="../_images/tutorials_CrossChatH_tutorial_8_2.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_8_3.png" src="../_images/tutorials_CrossChatH_tutorial_8_3.png" />
</div>
</div>
</section>
<section id="Find-specific-LR-pairs">
<h2>Find specific LR pairs<a class="headerlink" href="#Find-specific-LR-pairs" title="Permalink to this heading">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Table of top 20 most specific ligand-receptor interactions</span>
<span class="n">top20_specific_LR</span> <span class="o">=</span> <span class="n">ccH_obj</span><span class="o">.</span><span class="n">Detect_specific_LRs</span><span class="p">(</span><span class="n">topN</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">gene_ls</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CCL3&#39;</span><span class="p">,</span><span class="s1">&#39;CCR1&#39;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_ls</span><span class="p">:</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">ccH_obj</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;GnBu&#39;</span><span class="p">,</span><span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span><span class="n">vmax</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_11_0.png" src="../_images/tutorials_CrossChatH_tutorial_11_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_11_1.png" src="../_images/tutorials_CrossChatH_tutorial_11_1.png" />
</div>
</div>
</section>
<section id="Draw-CCC-of-specific-LR-pairs">
<h2>Draw CCC of specific LR pairs<a class="headerlink" href="#Draw-CCC-of-specific-LR-pairs" title="Permalink to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">ccH_obj</span><span class="o">.</span><span class="n">Draw_CCC_LR</span><span class="p">(</span><span class="n">ligand</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CCL3&#39;</span><span class="p">],</span><span class="n">receptor</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CCR1&#39;</span><span class="p">],</span><span class="n">CCC_threshold</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_13_0.png" src="../_images/tutorials_CrossChatH_tutorial_13_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">celltype_annotations</span> <span class="o">=</span> <span class="n">ccH_obj</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;annotations&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">jaccard_dist</span><span class="p">(</span><span class="n">ccH_obj</span><span class="o">.</span><span class="n">L_allresults</span><span class="p">[</span><span class="s1">&#39;selected_comm_ids&#39;</span><span class="p">],</span><span class="n">celltype_annotations</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_14_0.png" src="../_images/tutorials_CrossChatH_tutorial_14_0.png" />
</div>
</div>
</section>
<section id="Cluster-specific-LR-pairs">
<h2>Cluster specific LR pairs<a class="headerlink" href="#Cluster-specific-LR-pairs" title="Permalink to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">LR_ls</span> <span class="o">=</span> <span class="p">[[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">top20_specific_LR</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">20</span><span class="p">][[</span><span class="s1">&#39;Ligand&#39;</span><span class="p">,</span><span class="s1">&#39;Receptor&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">Cluster_LRs</span><span class="p">(</span><span class="n">LR_ls</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_16_0.png" src="../_images/tutorials_CrossChatH_tutorial_16_0.png" />
</div>
</div>
</section>
<section id="Draw-CCC-of-pathways">
<h2>Draw CCC of pathways<a class="headerlink" href="#Draw-CCC-of-pathways" title="Permalink to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">active_pathways</span> <span class="o">=</span> <span class="n">ccH_obj</span><span class="o">.</span><span class="n">Detect_active_pathways</span><span class="p">()</span>
<span class="n">pathway</span> <span class="o">=</span> <span class="s1">&#39;GRN&#39;</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">Draw_CCC</span><span class="p">(</span><span class="n">pathway</span><span class="p">,</span><span class="n">CCC_threshold</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_18_0.png" src="../_images/tutorials_CrossChatH_tutorial_18_0.png" />
</div>
</div>
</section>
<section id="Cluster-pathways">
<h2>Cluster pathways<a class="headerlink" href="#Cluster-pathways" title="Permalink to this heading">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">ccH_obj</span><span class="o">.</span><span class="n">Cluster_pathways</span><span class="p">(</span><span class="n">nclusters</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_20_0.png" src="../_images/tutorials_CrossChatH_tutorial_20_0.png" />
</div>
</div>
</section>
<section id="CrossChatH-on-spatial-data">
<h2>CrossChatH on spatial data<a class="headerlink" href="#CrossChatH-on-spatial-data" title="Permalink to this heading">¶</a></h2>
<p>Once we detected multiscale clusters using spatial information, the rest of procedures are similar to those applied to scRNA-seq datasets.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython2 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;../src/crosschat/data/mouse_wounded_skin_PO7_visium.h5ad&#39;</span><span class="p">)</span>
<span class="n">ccH_obj</span> <span class="o">=</span> <span class="n">cc</span><span class="o">.</span><span class="n">CrossChatH</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">species</span><span class="o">=</span><span class="s2">&quot;mouse&quot;</span><span class="p">)</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">prepare_adata</span><span class="p">(</span><span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="nb">input</span><span class="o">=</span><span class="s1">&#39;allgenes&#39;</span><span class="p">)</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">Multsicale_clustering_spatial</span><span class="p">(</span><span class="n">cluster_by</span><span class="o">=</span><span class="s2">&quot;allgenes&quot;</span><span class="p">)</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">select_partitions</span><span class="p">(</span><span class="n">max_nvi</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">window_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">basin_radius</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">lr</span> <span class="o">=</span> <span class="s2">&quot;L&quot;</span><span class="p">)</span>
<span class="n">ccH_obj</span><span class="o">.</span><span class="n">Draw_multiscale_umap</span><span class="p">(</span><span class="s2">&quot;L&quot;</span><span class="p">,</span><span class="n">spatial</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 100/100 [00:02&lt;00:00, 38.28it/s]
100%|██████████| 100/100 [01:28&lt;00:00,  1.13it/s]
100%|██████████| 100/100 [00:16&lt;00:00,  5.92it/s]
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/threadpoolctl.py:1019: RuntimeWarning: libc not found. The ctypes module in Python 3.8 is maybe too old for this OS.
  warnings.warn(
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/threadpoolctl.py:1019: RuntimeWarning: libc not found. The ctypes module in Python 3.8 is maybe too old for this OS.
  warnings.warn(
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/threadpoolctl.py:1019: RuntimeWarning: libc not found. The ctypes module in Python 3.8 is maybe too old for this OS.
  warnings.warn(
/Users/xinyiwang/.pyenv/versions/3.8.9/lib/python3.8/site-packages/threadpoolctl.py:1019: RuntimeWarning: libc not found. The ctypes module in Python 3.8 is maybe too old for this OS.
  warnings.warn(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_22_1.png" src="../_images/tutorials_CrossChatH_tutorial_22_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_22_2.png" src="../_images/tutorials_CrossChatH_tutorial_22_2.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_22_3.png" src="../_images/tutorials_CrossChatH_tutorial_22_3.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_CrossChatH_tutorial_22_4.png" src="../_images/tutorials_CrossChatH_tutorial_22_4.png" />
</div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">CrossChat</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../README.html"><em>CrossChat</em></a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Part I: CrossChatH</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Load-required-packages">Load required packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Load-data-and-preprocess">Load data and preprocess</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Perform-Multiscale-clustering">Perform Multiscale clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Draw-clustering-results">Draw clustering results</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Find-specific-LR-pairs">Find specific LR pairs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Draw-CCC-of-specific-LR-pairs">Draw CCC of specific LR pairs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Cluster-specific-LR-pairs">Cluster specific LR pairs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Draw-CCC-of-pathways">Draw CCC of pathways</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Cluster-pathways">Cluster pathways</a></li>
<li class="toctree-l2"><a class="reference internal" href="#CrossChatH-on-spatial-data">CrossChatH on spatial data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="CrossChatT_tutorial.html">Part II: CrossChatT</a></li>
<li class="toctree-l1"><a class="reference internal" href="CrossChatT_tutorial.html#CrossChatT-on-spatial-dataset">CrossChatT on spatial dataset</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">&lt;no title&gt;</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">&lt;no title&gt;</a></li>
      <li>Next: <a href="CrossChatT_tutorial.html" title="next chapter">Part II: CrossChatT</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2024, Xinyi Wang.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.0.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="../_sources/tutorials/CrossChatH_tutorial.ipynb.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>